lib/xebow/settings.ex

```diff
@@ -0,0 +1,85 @@
+defmodule Xebow.Settings do
+  @moduledoc """
+  Application settings that are persisted to disk.
+  """
+
+  alias RGBMatrix.Animation
+
+  require Logger
+
+  @settings_path Application.compile_env!(:xebow, :settings_path)
+
+  @animations_path Path.join(@settings_path, "animations.json")
+
+  @animations_version 1
+
+  @doc """
+  Create the directory for the settings files. This function is idempotent.
+  """
+  @spec create_dir!() :: :ok
+  def create_dir! do
+    File.mkdir_p!(@settings_path)
+  end
+
+  @doc """
+  Save the active animations to disk.
+  """
+  @spec save_active_animations!(animation_types :: [Animation.type()]) :: :ok
+  def save_active_animations!(animation_types) do
+    create_dir!()
+
+    payload =
+      Jason.encode_to_iodata!(%{
+        schema_version: @animations_version,
+        active_animation_types: animation_types
+      })
+
+    File.write!(@animations_path, payload)
+  end
+
+  @doc """
+  Load the active animations from disk.
+  """
+  @spec load_active_animations() ::
+          {:ok, [Animation.type()]}
+          | {:error, File.posix()}
+          | {:error, :schema_mismatch}
+  def load_active_animations do
+    with {:ok, payload} <- File.read(@animations_path),
+         animation_settings <- Jason.decode!(payload),
+         :ok <- validate_schema(animation_settings, @animations_version) do
+      animation_types =
+        animation_settings["active_animation_types"]
+        |> Enum.reduce([], fn animation_type, acc ->
+          try do
+            module = String.to_existing_atom(animation_type)
+            acc ++ [module]
+          rescue
+            ArgumentError ->
+              Logger.warn("Ignoring invalid animation setting: #{animation_type}")
+              acc
+          end
+        end)
```

Comment:
I left a reply to a similar comment earlier: https://github.com/ElixirSeattle/xebow/pull/109#discussion_r467698929

I tried another way of writing this, except rather than only extracting the anonymous function, I rewrote it in a way I think would be more idiomatic for having a function in Elixir do a reduce operation:

```ex
def load_active_animations do
  with {:ok, payload} <- File.read(@animations_path),
       animation_settings <- Jason.decode!(payload),
       :ok <- validate_schema(animation_settings, @animations_version) do
    animation_types =
      animation_settings["active_animation_types"]
      |> cast_animation_types()

    {:ok, animation_types}
  else
    error -> error
  end
end

defp cast_animation_types(animation_types),
  do: cast_animation_types(animation_types, [])

defp cast_animation_types([], acc),
  do: acc

defp cast_animation_types([animation_type | animation_types], acc) do
  acc =
    try do
      module = String.to_existing_atom(animation_type)
      acc ++ [module]
    rescue
      ArgumentError ->
        Logger.warn("Ignoring invalid animation setting: #{animation_type}")
        acc
    end

  cast_animation_types(animation_types, acc)
end
```

This came out to 33 lines. The former solution was 22 lines. I'm also not fond of how this is spreading out the logic of what was formerly a concise function. What mostly bothered me in the original implementation was the `try/reduce` cluttering things up, but a good way around that hasn't come to mind yet.

Reply:
There may be another interesting variant of this with the singular form of the function like you mentioned, except I'm not sure the accumulator is an appropriate input for the function. So what I'm thinking of is `cast_animation_type(animation_type)`. `Enum.reduce` would still need to sort out the accumulator, but this may hide the complex `try/rescue` stuff better. I'll have to try it out later.

Reply:
Something like

```elixir
#
animation_types =
  animation_settings["active_animation_types"]
  |> Enum.flat_map(&cast_animation_type/1)
#
defp cast_animation_type(animation_type) do
  try do
    [String.to_existing_atom(animation_type)]
  rescue
    ArgumentError ->
      Logger.warn("Ignoring invalid animation setting: #{animation_type}")
      []
  end
end
#
```

Is that sort of like what you're thinking about?

Reply:
:+1: yep! And good call on `capture_log`; I was able to turn the warnings back on in test.

Updated in 0655365.